hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -0,0 +1,447 @@
+package is.hail.expr.ir
+
+import is.hail.expr.types.physical._
+import is.hail.expr.types.virtual.{TArray, TTuple}
+import is.hail.utils._
+
+// Env contains all of the symbols values to reference
+// at root, the env is empty
+// In body of say tablemaprows, you are allowed to reference row, or global, etc
+// the names of row and global are bound in the body of tablemaprows
+// so tablemaprows appends to the environment
+// It's possible for a name to be bound twice
+// we may have nested array maps, and say each binds i
+// inside the inner map body, any reference to the name
+// FIXME: strip all requiredness logic when possible
+// TODO: We don't implicitly walk bottom-up for virtualTypes (not calling .typ) in all cases
+// TODO: If virtual type inference doesn't occur before this step, there may be subtle
+// TODO: behavior differences between IR Nodes (those that need .typ here and those that don't)
+// TODO: In cases where env is sent as is to child nodes, do we need copy? I believe no, as long as copy-on-write
+
+// TODO: aggregators: 1) use bindAgg/bindScan instead of bindEval? 2) proper passing of env, seems little needed
+object InferPType {
+  def apply(ir: IR, env: BindingEnv[PType]): PType = {
```

Comment:
High level --

I think the signature of this method should be 
```
def apply(ir: IR, env: BindingEnv[PType]): Unit = 
```

The first lines should probably be 

```
assert(ir._ptype == null)
ir._ptype = ir match { 
 ...
```

After the function exits, every node in the graph ending at `ir` should have a defined ptype. There are some places below (e.g. MakeArray) that don't fully infer the graph.

Reply:
Do we want to make `_ptype` a public property? We're currently using it as an internal check that this IR, has had InferPType called.

Regarding assert:
1) We already have an assert on the public method, ir.pType. I don't understand the need for 2 asserts, because the first assert is gatekeeper to inferPType; 2nd assert should never fail, unless we allow multipass inference
2) Regarding multipass inference: I could imagine this may be possible with circular dependencies between children. I'm not sure if this is allowed, I don't think so, but could it be in the future?

Parent -> Child1 -> binds Child2
Parent -> Child2 -> binds Child1

Reply:
> we want to make _ptype a public property

The `ir.inferSetPType(env)` model seems a bit weird to me, so probably making the _ptype public (or protected[ir]) is the only solution

Reply:
> 2nd assert should never fail

Good reason for an assert :)

Reply:
But it affects performance without benefit! I'll add in

Reply:
> > we want to make _ptype a public property
> 
> The `ir.inferSetPType(env)` model seems a bit weird to me, so probably making the _ptype public (or protected[ir]) is the only solution

I did this to avoid having implicit setter, which .typ did. Namely in InferType, any call to .typ also triggers InferType, which I think increases the chance of bugs, complex behavior.

Reply:
> But it affects performance without benefit! I'll add in

Assertions do have a benefit -- they verify that our assumptions about the state and usage of the code are actually true. Many of the bugs people report are caught by assertions, and that's way better than crashing somewhere more difficult to debug, or returning incorrect results.

You are right, though, that we need to weigh this benefit against the runtime cost. In this case, that cost is really small -- a fast check done once per IR node.

Reply:
> I did this to avoid having implicit setter, which .typ did.

I think `pType` should be:

```scala
def pType: PType = {
    assert(_ptype != null)
    ptype
}
```